### PR TITLE
Foorm Editor: Add Save Button

### PR DESCRIPTION
**Pull request by molly-moen**
*Note*: This PR is best viewed with whitespace changes hidden since I did some refactoring (in the file viewer go to settings gear->check hide whitespace changes).

Add save button to Foorm Editor and do some cleanup of the code/editor.

The save button is modeled after the Levelbuilder [SaveBar](https://github.com/code-dot-org/code-dot-org/blob/staging-next/apps/src/lib/levelbuilder/SaveBar.jsx). I made a different version instead of re-using the component because we didn't need a save and exit option, and to keep all the save logic together (including a warning pop-up when editing a published survey and parsing errors).

To enable saving easily this PR also adds referencing the forms by id instead of name/version which we kept stable in [this PR](https://github.com/code-dot-org/code-dot-org/pull/37887).

The save button is also only enabled for existing surveys--saving new surveys will be done as a follow up.

Minor improvements to the editor included in this PR:
- move the validation up above the editor (below live preview toggle)
- move the variables tab from next to the preview tab to next to the editor tab
- refactor large editor file into smaller files to make code easier to manage
- only call fill in library questions at most once per second (use debounce to achieve this)

The top of the editor now looks like this:
Screen Shot 2020-12-09 at 11 15 16 AM

Save warning message:
Screen Shot 2020-12-09 at 12 39 13 PM

Save bar with last saved timestamp:
Screen Shot 2020-12-09 at 11 12 56 AM

Save bar with error message:
Screen Shot 2020-12-09 at 11 15 00 AM

### Future work
- The How To linked from the editor still needs to be written. 
- Add save as new version, publish and save a new survey options.
- Add a warning if the user tries to leave the page with unsaved changes.

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLC-1047)

## Testing story
Tested locally with levelbuilder mode enabled. Tested that you cannot save an invalid survey, and saving is only allowed for levelbuilder users in levelbuilder mode.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/pd/foorm/editor/FoormEditorHeader.jsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
+// Metadata and buttons for the Foorm Editor. Shows form name and version,
```
Comment (molly-moen) [inline apps/src/code-studio/pd/foorm/editor/FoormEditorHeader.jsx:1]: nothing in this file is new--content moved from FoormEditor and validate button moved below live preview

File: apps/src/code-studio/pd/foorm/editor/FoormEditorPreview.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
+// Preview panel for Foorm editor.
```
Comment (molly-moen) [inline apps/src/code-studio/pd/foorm/editor/FoormEditorPreview.jsx:1]: this file's contents are also not new--just refactored from large FoormEditor file

File: apps/src/code-studio/pd/foorm/editor/FoormEditorPreview.jsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  },
+  errorMessage: {
+    fontWeight: 'bold',
+    padding: '1em'
+  }
+};
+
+class FoormEditorPreview extends Component {
+  static propTypes = {
+    formHasError: PropTypes.bool,
```
Comment (dmcavoy) [inline apps/src/code-studio/pd/foorm/editor/FoormEditorPreview.jsx]: maybe pull formHasError to the bottom and mark as from redux like you did in other places?

File: dashboard/app/controllers/foorm/forms_controller.rb [hunk 4]
```diff
@@ -9,12 +9,12 @@ def update_questions
       questions_json = get_questions
       published_state = questions_json['published']
 
-      if published_state.nil_or_empty?
```
Comment (dmcavoy) [inline dashboard/app/controllers/foorm/forms_controller.rb:12]: Why did these things in this method change from `nil_or_empty`? and `to_bool` ?

File: apps/src/code-studio/pd/foorm/editor/FoormEditorPreview.jsx (hunk 3 above)
Comment (molly-moen) [reply apps/src/code-studio/pd/foorm/editor/FoormEditorPreview.jsx]: good catch

File: dashboard/app/controllers/foorm/forms_controller.rb (hunk 4 above)
Comment (molly-moen) [reply dashboard/app/controllers/foorm/forms_controller.rb:12]: I had a bug here--I assumed the published state would come in as a string but it comes in as a boolean, so the extra checks weren't needed

**Review by dmcavoy — APPROVED** [review]
Molly the improvements you are making in this PR are great! I'm excited about the save button and the organization of the page. It really is looking so great!

I've looked through the changes as best I could and I think things look good. I have to say it was a little tricky to review. I run into this all the time myself where I get so excited about the changes I'm making I don't realize how big the change I'm making is. I think this PR could have been broken into a couple different PRs possibly to make reviewing easier. A couple ways you could have done this:
- Do the file reorganization before adding new features
- Have the save button be its own PR
- Move the features around into new locations in own PR

These are just some ideas and again I have this happen to me all the time and Dave is always giving me this feedback.  No need to do any of this now I think but definitely something to consider for the future.

**Comment by molly-moen** [comment]
@dmcavoy thanks! totally agree, it got too big.

**Pull request merged**